### PR TITLE
fix(engine): keep a Python thread's own name in the engine and in Python

### LINKED ISSUE #2368: Python-started threads show up as "External" in the profiler and engine logs
## Description

A thread that Python starts and that then calls into the engine loses its name: Python renames it `External`, and the engine calls it `External` too. In a pipeline subprocess these are the asyncio workers `DataConn` hands every pipe open, write and close to (`asyncio.to_thread`). So the profiler lists the threads that did the work as several identical `External` rows, and engine log lines from them cannot be matched to Python ones.

How it happens:

1. The worker (`asyncio_0`) calls into the engine, e.g. `pipe.writeText(...)`. The engine has no thread context for a thread it did not start, so the first `ThreadApi::thisCtx()` creates one with the default name `"External"` (`engine-core/apLib/async/ThreadApi.hpp:81`).
2. When the engine calls a Python node on the same thread, `setupDebug()` (`engine-lib/engLib/python/init.cpp:353-379`) copies the engine's name into `threading.current_thread().name`, overwriting `asyncio_0`.
3. The profiler resolves thread names from `threading._active`, so it reports `External`.

Develop baseline: `11800744d`.

## Steps to Reproduce

1. Start a pipeline that runs a Python node, e.g. webhook → parse → preprocessor_langchain → response, as task `DEMO`.
2. Profile the task while documents go through it:

```python
await client.cprofile_start(target='DEMO')
# upload a few documents to DEMO
await client.cprofile_stop(target='DEMO')
print(await client.cprofile_threads(target='DEMO'))
```

## Expected vs Actual Behavior

Expected: every thread keeps the name its creator gave it, `asyncio_0` … `asyncio_4`, in Python and in the engine.

Actual:

```
ID  NAME               TID     TIME  SHARE
 0  GlobalEventLoop  22052  47.124s  28.9%
 4  External         94680  23.399s  14.4%
 5  asyncio_4        69756  23.281s  14.3%
 2  External          5728  23.086s  14.2%
 3  External         69108  23.084s  14.2%
 1  External          2800  23.080s  14.2%
```

`asyncio_4` kept its name only because it never ran a Python node through the engine.

## Severity

P3 - Low (minor inconvenience): thread names in the profiler, engine logs and debugger; no functional impact.

## Platform

- [x] Windows
- [x] macOS
- [x] Linux

Reproduced on Windows; the code path is platform-independent.

## Version / Environment

develop `11800744d`, engine built from source, Windows 11.

### PR DESCRIPTION
**Pull request by asclearuc**
## Summary

- A thread Python starts (the asyncio workers the data path hands every pipe open, write and close to) was renamed `External` in Python on its first call back from the engine, and was `External` in the engine as well. So the profiler listed the threads that did the work as identical `External` rows.
- New `syncThreadName()` settles one name per thread on both sides. A thread Python started keeps its name and the engine takes it. A thread Python knows only as a `_DummyThread` takes the engine's name, as before, so engine threads, the main thread (`Main`) and foreign native threads are unchanged. It runs where Python enters the engine (`UnlockPython`) and where the engine enters Python (`setupDebug`).
- `UnlockPython` now holds `gil_scoped_release` as a member initialized after the name sync: a base class would release the GIL before any `UnlockPython` code runs. The placeholder name becomes `ThreadApi::ExternalName`.

## Type

fix

## Testing

- [x] Tests added or updated
- [x] Tested locally
- [ ] `./builder test` passes

New `engine-lib/test/python/thread_names.cpp` (engtest):

| Case | Thread | Expected (engine / Python) |
|---|---|---|
| `python_thread` | `threading.Thread(name="rr-py-thread")` entering the engine through `UnlockPython`, then a Python call through `callPython` | `rr-py-thread` / `rr-py-thread` |
| `python_thread_engine_first` | same, with an engine thread context made before `UnlockPython` | `rr-py-engine-first` / `rr-py-engine-first` |
| `engine_thread` | engine `async::Thread{"rr-engine"}` | `rr-engine` / `rr-engine` |
| `foreign_thread` | bare `std::thread` | `External` / `External` |

Before the change the first two failed (`"External" == "rr-py-thread"`); the last two passed and guard current behavior. After it, run through builder (`server:run-engtest`, `server:run-aptest`) on Windows: engtest "All tests passed (524 assertions in 32 test cases)", aptest "All tests passed (40742748 assertions in 101 test cases)".

Manual check on the rebuilt engine: a webhook → parse → preprocessor_langchain pipeline profiled over six documents lists its threads as `GlobalEventLoop` and `asyncio_0` … `asyncio_4` instead of `External`. The names are the same in a second session, where the threads existed before it started.

`./builder test` was not run locally.

## Checklist

- [x] Commit messages follow [conventional commits](https://www.conventionalcommits.org/)
- [x] No secrets or credentials included
- [ ] Wiki updated (if applicable)
- [ ] Breaking changes documented (if applicable)

## Linked Issue

Fixes #2368

## Summary by CodeRabbit

* **New Features**
  * Improved thread-name synchronization between the engine and Python for more consistent names across both environments.
  * Python-created and engine-created threads now appropriately retain or propagate their names.
  * Foreign threads continue to use the `External` name consistently.

* **Tests**
  * Added coverage for thread-name propagation across Python-created, engine-created, and foreign threads.

### CODE REVIEW

**Review by joshuadarron — DISMISSED** [review]
**Approving.** This touches the GIL and a hot path (`UnlockPython` runs on every pipe open/write/close), so I checked the parts that could bite rather than just the naming logic.

**The base → member change on `UnlockPython` is safe**

- The stated reason is correct C++: a base subobject initializes before any member or ctor body, so `gil_scoped_release` as a base would have dropped the GIL before `syncThreadName()` could touch Python. Moving it to a member declared *after* `m_syncName` fixes the ordering.
- Destruction still works out: members destruct in reverse declaration order, so `m_release` re-acquires the GIL first, and the dtor body still runs GIL-released exactly as it did when the release was a base. No behavioural change there.
- I checked all 56 `UnlockPython` sites across `packages/`. Every one is a plain RAII local (`UnlockPython unlock;`) — no `disarm()`, no member access, no slicing to `py::gil_scoped_release&`, no polymorphic or by-value use. Dropping the public base breaks nothing.
- The GIL precondition for the new Python calls in the ctor is the one that already existed: `gil_scoped_release` requires the GIL held anyway. And `syncThreadName()` is `noexcept` with a catch-all, so a `py::error_already_set` from `import("threading")` can't escape into a ctor on the data path.
- Hot-path cost is one thread-local bool test after the first call. Setting `tls_thread_named = true` before the work also makes the guard re-entrancy-safe.

**The naming logic is right**

- No dangling view. `RunCtx::setName` (`RunCtx.hpp:51`) `_copyTo`s into a fixed `Array<char,127>`, so passing a `TextView` over the local `std::string` at `init.cpp:392` copies. Same for `thisCtx(name)` creating the context.
- It updates the name that actually gets read: `getCurrentThreadName()` is `thisCtx()->name()` (`api.hpp:151`), which is what the logs, the profiler and `Backtrace.cpp:70` resolve. And `RunCtx::setName` is already a documented mid-flight operation (`Thread::setName` is public API), so the lock-free partial-name tolerance isn't being newly exercised.
- The main thread is preserved by the fall-through, which is subtler than it looks and worth calling out: engine ctx is `Main`, Python is `MainThread`, so `ctx->name() != ExternalName` skips the `setName`, the equality check then fails, and control drops to the dummy branch that renames Python to `Main`. Exactly the pre-change behaviour, without a special case for it.
- `setupProfiler`'s documented "MUST stay after `setupDebug()`" constraint (`call.hpp:187`) still holds — `setupDebug` calls `syncThreadName` first, and the new `UnlockPython` call only settles the name earlier.
- Nothing in the tree branches on the placeholder: `grep '"External"'` across `packages/` returns only the new constant and the test's assertions. Introducing `ThreadApi::ExternalName` doesn't strand a string comparison anywhere.

**Tests**

`rocketride_add_executable(engtest *.cpp *.hpp *.h ...)` globs, so `thread_names.cpp` is picked up with no CMake edit — and the named-namespace comment correctly anticipates the unity build sharing a TU with `profiler.cpp`. The bounded waits and "workers record, Catch2 thread asserts" discipline match `profiler.cpp`'s conventions.

On the unchecked `./builder test` box: CI runs `./builder test --verbose --sequential` (`_build.yaml:298`), and the `test` task includes `server:run-engtest` (`packages/server/scripts/tasks.js:1226`). Build is green on **Ubuntu, Windows and macOS**, so the four new cases ran on all three platforms, not just your local Windows run. That box can be ticked.

**Two nits, neither blocking**

1. `threading._DummyThread` (`init.cpp:384`) is private CPython API. It's the pragmatic check and it's still there in 3.13, but if it ever moves, the `AttributeError` is swallowed by the catch-all and — because `tls_thread_named` is already set — the thread is never named on either side again, with only a log line to say so. Catching that case and leaving the guard unset so a later call retries would make the degradation recoverable rather than permanent.

2. The sync updates the context name but not the OS-level thread name (`ap::async::setCurrentThreadName`), which is what a debugger, `top -H` and Windows minidumps show. Engine threads get both via `ThreadApi::threadEntry`; a Python-started thread now gets only the context one. Not a regression — it was `External` before and is correct now — but a one-line addition would make the debugger agree with the profiler. Fine as a follow-up.

Edge cases I checked and found benign: a Python thread genuinely named `External` still converges correctly; a name over 126 chars truncates in the `Name` array, fails the equality check and falls through to renaming the Python side to the truncated name — which still leaves both sides agreeing. No doc update is owed here; this is engine-internal and no doc records thread names.

File: packages/server/engine-lib/test/python/thread_names.cpp [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+import engtest_thread_names as _rr_thread_names
+
+
+def rr_thread_names_run(name, engine_first):
+    seen = []
+    thread = _rr_threading.Thread(
+        name=name,
+        target=lambda: seen.append(_rr_thread_names.enter_engine(engine_first)))
+    thread.start()
+    thread.join(60)
```
[the remark this answers is not in the document]
Comment (asclearuc) [reply packages/server/engine-lib/test/python/thread_names.cpp:155]: @coderabbit The worker is now started with `daemon=True` (ce0f2d09), so interpreter shutdown no longer waits for it: `threading._shutdown()` joins non-daemon threads only ("Wait until the Python thread state of all non-daemon threads get deleted"). `join(60)` stays as the bounded wait for the result, and `daemon` does not affect joining.

I did not move the probe into a child process. The hang you describe comes from finalization waiting for a non-daemon thread, and the flag removes exactly that. Process isolation would additionally cover a worker wedged *while holding the GIL*, where `Py_FinalizeEx()` would block on the GIL itself — but that case means the GIL seam under test is broken, every wait in this file is already bounded (60 s), and the failure path abandons the worker instead of joining it. A second process plus result plumbing and parent-side termination did not look worth it for that.

The other finding in this review (stack-owned `Names`/`Done` captured by reference while a timed-out worker keeps running) is fixed as suggested: both now live in one `State` behind a `shared_ptr` that each worker lambda captures by value.

Does that address it, or do you see a remaining path where the run can hang?

[the remark this answers is not in the document]
Comment (asclearuc) [reply packages/server/engine-lib/test/python/thread_names.cpp:155]: @coderabbit Agreed on the mechanics: `daemon=True` does not change GIL ownership, so a worker wedged inside `enter_engine()` while holding the GIL would still block `Py_FinalizeEx()` in `main.cpp`, after the timeout has already been reported. We are accepting that residual path in this PR, for four reasons.

1. It requires the seam under test to be broken. A worker can only hold the GIL forever if `LockPython`/`UnlockPython` or the name sync deadlocks — the very defect these cases exist to catch. The run then fails loudly; what is lost at the end is the summary line and the exit code, and CI ends the job on its own timeout with the failure already in the log.
2. This PR adds no new exposure. The same pattern predates it in the same binary: `python/profiler.cpp` abandons its workers on timeout (`thread.detach()` at lines 240, 291 and 366) and those workers also enter Python through `callPython`. Any of them stuck with the GIL hangs finalization identically. Fixing one of six sites with a child-process harness would leave the lane inconsistent.
3. Production does not depend on graceful finalization. A pipeline runs in its own engine subprocess, and the task server kills it rather than waiting: `engine.terminate()`, then `engine.kill()` after a timeout (`ai/modules/task/task_engine.py`), on top of the task's idle timer and TTL. There is no `Py_FinalizeEx()` path to block, so the hang you describe is test-harness-only.
4. Cost. A child process means spawning the test binary with a Catch2 filter, a marker for child mode, result plumbing and parent-side termination on both Windows and POSIX — worth it only if applied to the whole lane, which is a change to tests that are not part of this PR.

If the seam ever does break, the fix belongs in the lane as a whole rather than in this one file, and we will take it up there.

**Review by stepmikhaylov — APPROVED** [review]

**Review by dsapandora — APPROVED** [review]

File: packages/server/engine-lib/engLib/python/init.cpp [hunk 2]
```diff
@@ -350,32 +350,61 @@ void setupDebug() noexcept {
...
+            std::string name =
+                py::cast<std::string>(currentThread.attr("name"));
+
+            // A context made now is born with the name; one made earlier
+            // holds only the placeholder
+            auto ctx = ap::async::ThreadApi::thisCtx(name);
+            if (ctx->name() == ap::async::ThreadApi::ExternalName)
+                ctx->setName(ap::TextView{name});
+
+            if (ctx->name() == ap::TextView{name}) {
```
Comment (dsapandora) [inline packages/server/engine-lib/engLib/python/init.cpp:394]: `RunCtx::Name` is `Array<char, 127>`, so a Python name longer than 126 bytes gets truncated by `setName` and this comparison fails. It then falls through to line 403 and writes the truncated engine name back into Python, the overwrite this PR sets out to avoid. If the cut lands inside a UTF-8 sequence, the setter throws and the engine keeps an invalid UTF-8 name that ends up in logs.

Comparing against the truncated prefix (or not writing back when the thread isn't a `_DummyThread`), truncating on a code-point boundary, and a test with a long / non-ASCII name would cover it. Follow-up is fine since this is merged.
